### PR TITLE
Allow using existing socket

### PR DESCRIPTION
**Pull request by highwall**
To support one socket as both server and client.

### CODE REVIEW

**Comment by mcollina** [comment]
Can you please add unit tests for this?

**Comment by mcollina** [comment]
Great work!!

File: README.md [hunk 1]
```diff
@@ -494,6 +494,8 @@ Opts is an optional object with the following optional properties:
 * `type`: `'udp4'` or `'udp6'` if we want an Agent on an IPv4 or IPv6
   UDP socket.
 
+* `sock`: use existing sock instead of creating a new one.
+
```
Comment (mcollina) [inline README.md]: Can you please use socket instead of sock?

File: lib/agent.js [hunk 2]
```diff
@@ -111,6 +117,8 @@ Agent.prototype._doClose = function() {
   for (var k in this._msgIdToReq)
     this._msgIdToReq[k].sender.reset()
 
+  if (this._keep) return;
```
Comment (mcollina) [inline lib/agent.js]: Can you please remove this._keep, and move the check on opts.sock in this function?

File: lib/server.js [hunk 3]
```diff
@@ -255,6 +255,9 @@ function expiry(block2cache, k) {
 
 CoAPServer.prototype._handle = function(packet, rsinfo) {
 
+  // Make sure this is a request
+  if (packet.code[0] != '0') return;
+
```
Comment (mcollina) [inline lib/server.js]: why this?

File: README.md (hunk 1 above)
Comment (slowdone) [inline README.md]: OK, I will update.

File: lib/server.js (hunk 3 above)
Comment (slowdone) [inline lib/server.js]: If the code[0] is not 0, this packet is a response of a previous request from the same socket.

Comment (mcollina) [inline lib/server.js]: Ok. Could you add a bigger comment explaining this is needed to reuse the socket?

The problem with this change is when the socket is not reused, as the spec says we need to reply with something (couldn't remember now, please check), if we couldn't recognize the message.

Comment (highwall) [inline lib/server.js]: OK, I will do this later, for I'm really busy recently.

**Comment by mcollina** [comment]
Any update on this?

**Comment by radiumfl** [comment]
Dear Highwall and Mcollina, it's a wonderful feature. How to use it?
Thanks in advance! Radiumfl

**Comment by neophob** [comment]
+1

**Comment by highwall** [comment]
All test cases pass.

File: lib/agent.js [hunk 4]
```diff
@@ -44,19 +44,24 @@ function Agent(opts) {
   if (!opts.type)
     opts.type = 'udp4'
 
+  if (opts.socket) {
+    delete opts.port;
```
Comment (GiedriusM) [inline lib/agent.js]: Project code style [CS] does not use ";" as statement separator, so it should be removed

File: lib/agent.js [hunk 5]
```diff
@@ -44,19 +44,24 @@ function Agent(opts) {
...
 util.inherits(Agent, events.EventEmitter)
 
-Agent.prototype._init = function initSock() {
+Agent.prototype._init = function initSock(socket) {
   if (this._sock)
     return
 
   var that = this
-  this._sock = dgram.createSocket(this._opts.type, function(msg, rsinfo) {
+  this._sock = socket || dgram.createSocket(this._opts.type);
```
Comment (GiedriusM) [inline lib/agent.js]: [CS] Remove ";"

File: lib/agent.js [hunk 6]
```diff
@@ -70,6 +75,11 @@ Agent.prototype._init = function initSock() {
       return
     }
 
+    if (packet.code[0] === '0' && packet.code !== '0.00') {
+      // ignore this packet since it's not a response.
+      return;
```
Comment (GiedriusM) [inline lib/agent.js]: CS: Remove ";"

File: lib/agent.js [hunk 7]
```diff
@@ -111,6 +121,8 @@ Agent.prototype._doClose = function() {
   for (var k in this._msgIdToReq)
     this._msgIdToReq[k].sender.reset()
 
+  if (this._opts.socket) return;
```
Comment (GiedriusM) [inline lib/agent.js]: [CS] remove ";"
[CS] the return statement should be placed into a separate line

File: lib/server.js [hunk 8]
```diff
@@ -287,6 +287,12 @@ function expiry(block2cache, k) {
 
 CoAPServer.prototype._handle = function(packet, rsinfo) {
 
+  // Make sure this is a request
+  if (packet.code[0] !== '0') {
+    // just ignore this packet.
+    return;
```
Comment (GiedriusM) [inline lib/server.js]: [CS] remove ";"

File: test/common.js [hunk 9]
```diff
@@ -12,4 +12,3 @@ var portCounter = 9042
 global.nextPort = function() {
   return ++portCounter
 }
-
```
Comment (GiedriusM) [inline test/common.js]: This empty new line fix is not related to the issue, therefore it should go into a separate pull-request and reverted in this one.

File: test/share-socket.js [hunk 10]
```diff
@@ -0,0 +1,423 @@
...
+    , port
+
+  beforeEach(function(done) {
+    port = nextPort()
+    server = coap.createServer()
+    server.listen(port, function() {
+      coap.globalAgent = new coap.Agent({
+        socket: server._sock
+      });
+      done();
```
Comment (GiedriusM) [inline test/share-socket.js]: [CS] This and all the other ";" in this file should be removed

**Comment by highwall** [comment]
@GiedriusM modified

File: lib/server.js [hunk 11]
```diff
@@ -287,6 +287,11 @@ function expiry(block2cache, k) {
 
 CoAPServer.prototype._handle = function(packet, rsinfo) {
 
+  if (packet.code[0] !== '0') {
```
Comment (GiedriusM) [inline lib/server.js]: This section needs an explanation. Could you please add the following comment:
According to RFC7252 Section 4.2 receiving a confirmable messages that can't be processed, should be rejected by ignoring it AND sending a reset. In this case confirmable response message would be silently ignored, which is not exactly as stated in the standard. However, sending a reset would interfere with a coap client which is re-using a socket (see pull-request #131).

File: test/request.js [hunk 12]
```diff
@@ -81,6 +81,7 @@ describe('request', function() {
     var req = request('coap://aaa.eee:' + 1234)
 
     req.once('error', function () {
+      coap.globalAgent.abort(req);
```
Comment (GiedriusM) [inline test/request.js]: [CS] remove ";"

File: test/share-socket.js [hunk 13]
```diff
@@ -0,0 +1,423 @@
...
+      res.end()
+    })
+  })
+
+  describe('formats', function() {
+    var formats = [ 'text/plain', 'application/link-format',
+      'application/xml', 'application/octet-stream',
+      'application/exi', 'application/json', 'application/cbor' ]
+
+    ;['Accept', 'Content-Format'].forEach(function(option) {
```
Comment (GiedriusM) [inline test/share-socket.js]: [CS] remove ";"

File: test/share-socket.js [hunk 14]
```diff
@@ -0,0 +1,423 @@
...
+      , completed = 2
+
+    server.on('request', function(req, res) {
+      res.write('hello')
+      setTimeout(function() {
+        res.end('world')
+      }, 10)
+    })
+
+    ;[req1, req2].forEach(function(req) {
```
Comment (GiedriusM) [inline test/share-socket.js]: [CS] remove ";"

File: test/share-socket.js (hunk 13 above)
Comment (highwall) [inline test/share-socket.js]: Removing this `;` cause error. 
```Caught exception: TypeError: Cannot read property 'forEach' of undefined```
I think that's why test/end-to-end.js also keeps this `;`

File: test/share-socket.js (hunk 14 above)
Comment (highwall) [inline test/share-socket.js]: save as above

File: test/share-socket.js (hunk 13 above)
Comment (GiedriusM) [inline test/share-socket.js]: Ah sorry, I was too hasty to ctrl+f ; and comment, without looking at it closer :)
